File: backends/java/zappy-application/src/main/java/com/zappymart/application/accounts/ViewWishlist.java
package com.zappymart.application.accounts;

import com.zappymart.application.Visitor;
import com.zappymart.application.ports.ProductRepository;
import com.zappymart.application.ports.WishlistRepository;
import com.zappymart.domain.accounts.Wishlist;
import com.zappymart.domain.catalogue.Product;

import java.util.List;

public final class ViewWishlist {

    private final WishlistRepository wishlistRepository;
    private final ProductRepository productRepository;

    public ViewWishlist(WishlistRepository wishlistRepository, ProductRepository productRepository) {
        this.wishlistRepository = wishlistRepository;
        this.productRepository = productRepository;
    }

    public List<Product> execute(Visitor visitor) {
        return forOwner(WishlistOwner.of(visitor));
    }

    public List<Product> forOwner(String ownerId) {
        if (ownerId == null) {
            return List.of();
        }
        return productsOf(wishlistRepository.ofOwner(ownerId));
    }

    public List<Product> productsOf(Wishlist wishlist) {
        List<Product> saved = productRepository.byIds(wishlist.productIds());
        return wishlist.productIds().stream()
                .flatMap(productId -> saved.stream().filter(product -> product.id().equals(productId)))
                .toList();
    }
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/accounts/ViewWishlist.kt
package nl.zappymart.application.accounts

import nl.zappymart.application.Visitor
import nl.zappymart.application.ports.ProductRepository
import nl.zappymart.application.ports.WishlistRepository
import nl.zappymart.domain.catalogue.Product

class ViewWishlist(
    private val wishlists: WishlistRepository,
    private val products: ProductRepository,
    private val wishlistOwner: WishlistOwner,
) {

    fun execute(visitor: Visitor): List<Product> {
        val ownerId = wishlistOwner.forReading(visitor) ?: return emptyList()
        return inWishlistOrder(wishlists.findByOwnerId(ownerId).productIds)
    }

    fun inWishlistOrder(productIds: List<String>): List<Product> {
        if (productIds.isEmpty()) {
            return emptyList()
        }
        val found = products.findAllByIds(productIds).associateBy { product -> product.id }
        return productIds.mapNotNull { productId -> found[productId] }
    }
}
